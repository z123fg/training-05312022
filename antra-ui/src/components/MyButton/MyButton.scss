@use "../../scss/variable";

.btn{
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  border-radius: 4px;
}

$color-map: (
  primary:(
    default:variable.$primary-color,
    hover:#303F9F,
    disabled:rgba(0, 0, 0, 0.12)
  ),
  secondary:(
    default:variable.$secondary-color,
    hover:#C51162,
    disabled:rgba(0, 0, 0, 0.12)
  ),
  default:(
    default:variable.$default-color,
    hover:#D5D5D5,
    disabled:rgba(0, 0, 0, 0.12)
  )
);

@mixin color-variant-class($color, $variant){
  .btn-#{$color}-#{$variant}{
    @if $variant == "contained"{
      color:white;
      background-color: map-get(map-get($color-map,$color),"default");
      border:none;
      filter:drop-shadow(0px 3px 1px rgba(0, 0, 0, 0.2)) drop-shadow(0px 2px 2px rgba(0, 0, 0, 0.14)) drop-shadow(0px 1px 5px rgba(0, 0, 0, 0.12));
    }@else if $variant == "outlined"{
      background: none;
      color:map-get(map-get($color-map,$color),"default");
      border:1px solid rgba(map-get(map-get($color-map,$color),"default"), 0.5);
    }@else{
      border:none;
      background: none;
      color: map-get(map-get($color-map,$color),"default");
    }
  }
}

$colors:("primary","secondary","default");
$variants: ("contained", "outlined","text");

@mixin generate-color-variant-classes{
  @each $variant in $variants{
    @each $color in $colors{
      @include color-variant-class($color,$variant );
    }
  }
}

@include generate-color-variant-classes()

