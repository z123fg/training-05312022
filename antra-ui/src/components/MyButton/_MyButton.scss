@use "../../scss/variable";
@use "../../scss/mixin";


/* 
  variable
  mixin
  nested selector
  loop
  if else
  map
  list
  @use
  @include
*/

.btn {
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  border-radius: 4px;
  position:relative;
  overflow: hidden;
}

$color-map: (
  primary: (
    default: variable.$primary-color,
    hover: variable.$primary-color-hover,
    disabled: rgba(0, 0, 0, 0.12),
  ),
  secondary: (
    default: variable.$secondary-color,
    hover: variable.$secondary-color-hover,
    disabled: rgba(0, 0, 0, 0.12),
  ),
  default: (
    default: variable.$default-color,
    hover: variable.$default-color-hover,
    disabled: rgba(0, 0, 0, 0.12),
  ),
);

@mixin color-variant-class($color, $variant) {
  .btn-#{$color}-#{$variant} {
    @if $variant == "contained" {
      color: white;
      background-color: map-get(map-get($color-map, $color), "default");
      border: none;
      filter: drop-shadow(0px 3px 1px rgba(0, 0, 0, 0.2))
        drop-shadow(0px 2px 2px rgba(0, 0, 0, 0.14))
        drop-shadow(0px 1px 5px rgba(0, 0, 0, 0.12));
    } @else if $variant == "outlined" {
      background: none;
      color: map-get(map-get($color-map, $color), "default");
      border: 1px
        solid
        rgba(map-get(map-get($color-map, $color), "default"), 0.23);
    } @else {
      border: none;
      background: none;
      color: map-get(map-get($color-map, $color), "default");
    }
  }
  .btn-#{$color}-#{$variant}:hover {
    @if $variant == "contained" {
      background-color: map-get(map-get($color-map, $color), "hover");
      filter: none;
    } @else if $variant == "outlined" {
      background-color: rgba(
        map-get(map-get($color-map, $color), "default"),
        0.08
      );
      border: 1px
        solid
        rgba(map-get(map-get($color-map, $color), "default"), 0.23);
    } @else {
      background-color: rgba(
        map-get(map-get($color-map, $color), "default"),
        0.08
      );
    }
  }

  .btn-#{$color}-#{$variant}:disabled {
    @if $variant == "contained" {
      background-color: variable.$disabled-color;
      filter: none;
      color: variable.$disabled-font-color;
    } @else if $variant == "outlined" {
      background: none;
      color: variable.$disabled-font-color;
      border: 1px solid variable.$disabled-color;
    } @else {
      background: none;
      color: variable.$disabled-font-color;
    }
    pointer-events: none;
  }

  .btn-ripple-#{$color}-#{$variant}{
    pointer-events: none;
    animation: rippleEffect 0.6s linear;
    border-radius: 50%;
    @if $variant == "contained" {
      background-color: white;
    }@else{
      background-color: map-get(map-get($color-map, $color), "hover");
    }
  }
}



$colors: ("primary", "secondary", "default");
$variants: ("contained", "outlined", "text");

@mixin generate-color-variant-classes {
  @each $variant in $variants {
    @each $color in $colors {
      @include color-variant-class($color, $variant);
    }
  }
}

@include generate-color-variant-classes();

.btn-small {
  min-width: 97px;
  height: 30px;
  span {
    @include mixin.font-small;
    pointer-events: none;
  }
}

.btn-medium {
  min-width: 113px;
  height: 36px;
  span {
    @include mixin.font-medium;
    pointer-events: none;
  }
}

/* 
.btn-medium{
  min-width: 113px;
  height: 36px;
}

.btn-medium span{
    @include mixin.font-medium;
    pointer-events: none;
}


*/

.btn-large {
  min-width: 131px;
  height: 42px;
  span {
    @include mixin.font-large;
    pointer-events: none;
  }
}

.btn span{
  user-select: none;
}




@keyframes rippleEffect {
  0% {
    width:0px;
    height:0px;
    opacity: 0.3;
  }
  50%{
    width:75px;
    height:75px;
    opacity: 0.1;
  }
  100%{
    width:150px;
    height:150px;
    opacity: 0;
  }
}