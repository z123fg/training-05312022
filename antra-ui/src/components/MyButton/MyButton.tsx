import { FC, MouseEvent, ReactNode } from "react";
import "./MyButton.scss"

type ButtonColor = "primary" | "secondary" | "default";

type ButtonSize = "small" | "medium" | "large";

type ButtonVariant = "contained" | "outlined" | "text";


interface IMyButtonProps {
  color?: ButtonColor;
  size?: ButtonSize;
  variant?: ButtonVariant;
  disabled?: boolean;
  children?: ReactNode;
  onClick?:(event:MouseEvent) => void
}

const MyButton: FC<IMyButtonProps> = ({
  color = "primary",
  size = "medium",
  variant = "contained",
  disabled = false,
  children,
  onClick
}) => {

  const handleClick = (e:MouseEvent) => {
    onClick?.(e);
  }

  const constructClassName:()=>string = () => {
    const colorVariantCls = `btn-${color}-${variant}`
    return ["btn", colorVariantCls].join(" ")
  }

  //"btn btn-large"

  return (
    <button onClick={handleClick} className={constructClassName()}>
        <span>{children}</span>
    </button>
  )
}

export default MyButton