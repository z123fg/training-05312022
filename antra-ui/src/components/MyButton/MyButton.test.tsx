import { queryByText, findByText, render, screen, fireEvent, waitFor } from "@testing-library/react"
import React from "react"
import MyButton from "./MyButton"


//react testing library, enzyme
//test suite

//unit test
//behavior test, integration test, end to end(e2e) test: cypress, headless browser
//code coverage
describe("MyButton test cases", () => {
  test("snapshot test for MyButton", () => {
    const view = render(<MyButton>SUBMIT</MyButton>);
    expect(view.asFragment()).toMatchSnapshot();
  });

  test("MyButton component should be rendered", () => {
    render(<MyButton>SUBMIT</MyButton>);
    const buttonEl = screen.queryByText("SUBMIT");
    expect(buttonEl).toBeInTheDocument();
  });

  test("MyButton should change its color based on color props", async () => {
    const { rerender } = render(<MyButton color="primary">SUBMIT</MyButton>);
    let buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-primary-contained");
    rerender(<MyButton color="secondary">SUBMIT</MyButton>);
    buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-secondary-contained");
    rerender(<MyButton color="default">SUBMIT</MyButton>);
    buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-default-contained");
  })


  test("MyButton should change its variant based on variant props", async () => {
    const { rerender } = render(<MyButton variant="contained">SUBMIT</MyButton>);
    let buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-primary-contained");
    rerender(<MyButton variant="outlined">SUBMIT</MyButton>);
    buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-primary-outlined");
    rerender(<MyButton variant="text">SUBMIT</MyButton>);
    buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-primary-text");
  })

  test("MyButton should change its size based on size props", async () => {
    const { rerender } = render(<MyButton size="medium">SUBMIT</MyButton>);
    let buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-medium");
    rerender(<MyButton size="small">SUBMIT</MyButton>);
    buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-small");
    rerender(<MyButton size="large">SUBMIT</MyButton>);
    buttonEl = await screen.findByRole("button");
    expect(buttonEl).toHaveClass("btn-large");
  })

  test("MyButton should be disabled if the disabled props is true, should invoke onClick when it is not disabled", async () => {
    const mockClickHandler = jest.fn();
    const { rerender } = render(<MyButton onClick={mockClickHandler}>SUBMIT</MyButton>);
    let buttonEl = await screen.findByRole("button");
    fireEvent.click(buttonEl);
    expect(mockClickHandler).toBeCalledTimes(1);
    rerender(<MyButton onClick={mockClickHandler} disabled>SUBMIT</MyButton>);
    fireEvent.click(buttonEl);
    expect(mockClickHandler).toBeCalledTimes(1);
  });

  test("MyButton should have rippple animation on click", async () => {
    render(<MyButton>SUBMIT</MyButton>);
    let buttonEl = await screen.findByRole("button");
    let rippleEls = screen.queryAllByTestId("ripple-element");
    expect(rippleEls).toHaveLength(0);
    fireEvent.click(buttonEl);
    rippleEls = screen.queryAllByTestId("ripple-element");
    expect(rippleEls).toHaveLength(1);
    fireEvent.click(buttonEl);
    rippleEls = screen.queryAllByTestId("ripple-element");
    expect(rippleEls).toHaveLength(2);

    rippleEls.forEach(rippleEl => {
      fireEvent.animationEnd(rippleEl);
    });
    rippleEls = screen.queryAllByTestId("ripple-element");
    expect(rippleEls).toHaveLength(0);
  });
})