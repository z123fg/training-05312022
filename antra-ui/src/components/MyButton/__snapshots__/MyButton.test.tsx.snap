// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyButton test cases snapshot test for MyButton 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary-contained btn-medium"
    id="123"
  >
    <span>
      SUBMIT
    </span>
  </button>
</DocumentFragment>
`;
